### PR TITLE
:rotating_light: [`Flash Attention`] Fix sliding window size

### LINKED ISSUE #39408: Off-by-one error when using flash_attention with a sliding window
### System Info

Looking at https://github.com/huggingface/transformers/commit/6017f5e8ed33d48096cdf8630d1cc7cbf2550c90

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

From the [flash attention documentation](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_attn_interface.py#L1016) for the window_size argument:
```
    If window_size != (-1, -1), implements sliding window local attention. Query at position i
    will only attend to keys between [i - window_size[0], i + window_size[1]] inclusive.
```

Transformers [currently sets window_size size like so](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_flash_attention_utils.py#L489) when using flash attention :
```
flash_kwargs = {"window_size": (sliding_window, sliding_window)} if use_sliding_windows else {}
```

This means that from a given token, we will look backward by `sliding_window` and then forward by `sliding_window`, resulting in a total window size of `2 * sliding_window + 1`. If causal masking is applied, then it is equivalent to setting `(sliding_window, 0)` and the total window  size is `sliding_window + 1`.

### Expected behavior

The expected behavior is for the flash attention implementation of sliding window attention to match the other implementations, which currently use [this function](https://github.com/huggingface/transformers/blob/6017f5e8ed33d48096cdf8630d1cc7cbf2550c90/src/transformers/masking_utils.py#L76C1-L85C22) to apply a sliding window (note that transformer's flash attention usage explicitly bypasses modifying an attention mask with this function):
```
def sliding_window_overlay(sliding_window: int) -> Callable:
    """
    This is an overlay depicting a sliding window pattern. Add it on top of a causal mask for a proper sliding
    window mask.
    """

    def inner_mask(batch_idx: int, head_idx: int, q_idx: int, kv_idx: int) -> bool:
        return kv_idx > q_idx - sliding_window

    return inner_mask
```
When combined with a causal mask as intended, this function retains `sliding_window - 1` tokens to the left, plus the current token for a total window size of `sliding_window`.

Suggested fix:
```
flash_kwargs = {"window_size": (sliding_window - 1, sliding_window)} if use_sliding_windows else {}
```

### PR DESCRIPTION
**Pull request by vasqu**
Fixes #39408

This issue will occur especially during the prefill phase when sliding window < prefill size. Newer cache refactors automatically remove kvs beyond the window so inference would/should have been less affected. Small script to show the effects:
```python
import torch
from transformers import AutoTokenizer, AutoConfig, MistralForCausalLM


torch_device = "cuda"

config = AutoConfig.from_pretrained("mistralai/Mistral-7B-v0.1")
config.sliding_window = 4
model = MistralForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, config=config, device_map={"": torch_device}
)

prompt = "My favourite condiment is "
tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1", use_fast=False)
inputs = tokenizer(prompt, return_tensors="pt").to(torch_device)

# greedy generation outputs
generated_ids = model.generate(**inputs, max_new_tokens=20, do_sample=False)
text_sdpa = tokenizer.decode(generated_ids[0], skip_special_tokens=True)

model.set_attn_implementation("flash_attention_2")
generated_ids = model.generate(**inputs, max_new_tokens=20, do_sample=False)
text_fa = tokenizer.decode(generated_ids[0], skip_special_tokens=True)

print("sdpa")
print(text_sdpa)
print()
print("flash")
print(text_fa)
``` 
Outputs should match with the fix, before they were not. Will add a common test so it won't happen again. Edit: modified the existing test instead

cc @Cyrilvallez @ArthurZucker

### CODE REVIEW

**Review by tomaarsen — COMMENTED** [review]
As requested by @vasqu, I looked into whether we would also need to make changes on the ModernBERT side, as it's one of the few encoders with sliding window attention. Some details:

* ModernBERT (sadly) uses rather custom code for its SDPA, Eager, and FA2. We even had a PR for FlexAttention, but the performance was pretty meh. This means that any changes or issues in `modeling_flash_attention_utils.py` don't necessarily propagate to ModernBERT.
* The `ModernBertConfig` has its own parameter for the sliding window called `local_attention`: `The window size for local attention.` It defaults to 128.
* For the Eager and SDPA paths, a `sliding_window_mask` is created by masked-filling the 4d attention mask such that `distance <= self.config.local_attention // 2`. In practice, this results in attending to `local_attention // 2` tokens prior to token $i$, token $i$ itself, and `local_attention // 2` tokens after token $i$. So, `local_attention + 1` tokens are attended to, assuming that `local_attention` is divisible by 2 (which is essentially always).
* For the FA2 path, a `local_attention` tuple is created like `(config.local_attention // 2, config.local_attention // 2)`, which is passed directly as the `window_size` parameter to `flash_attn_varlen_qkvpacked_func` from FA2. Like is clear from this PR and its related issue, this also means attending to `local_attention // 2` tokens prior to token $i$, token $i$ itself, and `local_attention // 2` tokens after token $i$, for a total of `local_attention + 1` tokens.

In terms of conclusions, the FA2 and Eager/SDPA paths should be equivalent for ModernBERT (my understanding with this PR is that this might not be the case for most LLMs implementing FA2 via the `modeling_flash_attention_utils.py`?). 
Beyond that, ModernBERT defines its SWA a lot differently than the LLMs do, i.e. with `local_attention` instead of `sliding_window`. This parameter doesn't immediately imply that the window size is at most `local_attention`. 
Also, ModernBERT was trained with this `local_attention + 1` approach, I'm pretty sure. When I implemented the ModernBERT architecture into transformers, I was able to match the performance of the token embeddings of the FA2 route with 100% accuracy, I believe. That would not have been the case if the original implementation didn't also use `local_attention + 1`. 
And lastly, ModernBERT definitely doesn't have an issue where the total window size is accidentally twice as big (e.g. `2 * local_attention + 1`).

I don't really see a clear reason to make changes on the ModernBERT side, but you or the other transformers maintainers might disagree!

- Tom Aarsen

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -3493,6 +3493,10 @@ def flash_attn_inference_equivalence(self, attn_implementation: str, padding_sid
             # flash attention variants does not always support arbitrary headim
             config = self._prepare_config_headdim(config, 16)
 
+            # forcing sliding window to go over the prefill size to check for SWA correctness
+            if getattr(config, "sliding_window", None):
+                config.sliding_window = 2
```
Comment (vasqu) [inline tests/test_modeling_common.py]: Mistral would fail now w/o this fix

**Review by Cyrilvallez — APPROVED** [review]
Yes perfect! Thanks @tomaarsen for confirming that we do need both sides!

**Comment by Cyrilvallez** [comment]
What would be amazing @vasqu is if you can run slow tests locally for some main models using sliding window, such as Mistral/Gemma2,3, to see if some Integrations tests are actually impacted (it may not be the case at all). Let's try to anticipate a bit if we can, before seeing tomorrow's slow CI report! (make sure to compare to main as some tests are already failing)

**Comment by vasqu** [comment]
I checked for some important models (see list below) locally; note that the CI never installs FA which I also pointed out internally but we haven't gotten to it yet - wasn't possible on T4 runners. 

Checked models:
- Gemma 
  - [x] 2
  - [x] 3
  - [x] 3n 
    - See #39764 --> Integration tests seem to be skipped atm 
- Mistral
  - [x] 1
  - [ ] 3 
    - Can't use fa due to limitations in Pixtral

**Comment by vasqu** [comment]
In essence, it works as expected and only gemma3n potentially needs updates which should be done in the other PR

**Comment by Cyrilvallez** [comment]
Merging as it will help tests!
